src/trio/testing/_raises_group.py

```diff
@@ -332,61 +313,107 @@ class RaisesGroup(
+        exception: type[Ec2] | Matcher[Ec2],
+        *other_exceptions: type[Ec2] | Matcher[Ec2],
         match: str | Pattern[str] | None = None,
-        check: Callable[[BaseExceptionGroup[E]], bool] | None = None,
+        check: Callable[[ExceptionGroup[Ec2]], bool] | None = None,
+    ) -> None: ...
+
+    @overload
+    def __init__(
+        self: RaisesGroup[ExceptionGroup[Ec3]],
+        exception: RaisesGroup[Ec3],
+        *other_exceptions: RaisesGroup[Ec3],
+        match: str | Pattern[str] | None = None,
+        check: Callable[[ExceptionGroup[ExceptionGroup[Ec3]]], bool] | None = None,
     ) -> None: ...
 
+    @overload
     def __init__(
-        self,
-        exception: type[E] | Matcher[E] | E,
-        *other_exceptions: type[E] | Matcher[E] | E,
+        self: RaisesGroup[Ec2 | ExceptionGroup[Ec3]],
+        exception: type[Ec2] | Matcher[Ec2] | RaisesGroup[Ec3],
+        *other_exceptions: type[Ec2] | Matcher[Ec2] | RaisesGroup[Ec3],
+        match: str | Pattern[str] | None = None,
+        check: (
+            Callable[[ExceptionGroup[Ec2 | ExceptionGroup[Ec3]]], bool] | None
+        ) = None,
+    ) -> None: ...
+
+    # same as the above 3 but handling BaseException
+    @overload
+    def __init__(
+        self: RaisesGroup[E2],
+        exception: type[E2] | Matcher[E2],
+        *other_exceptions: type[E2] | Matcher[E2],
+        match: str | Pattern[str] | None = None,
+        check: Callable[[BaseExceptionGroup[E2]], bool] | None = None,
+    ) -> None: ...
+
+    @overload
+    def __init__(
+        self: RaisesGroup[BaseExceptionGroup[E3]],
+        exception: RaisesGroup[E3],
+        *other_exceptions: RaisesGroup[E3],
+        match: str | Pattern[str] | None = None,
+        check: (
+            Callable[[BaseExceptionGroup[BaseExceptionGroup[E3]]], bool] | None
+        ) = None,
+    ) -> None: ...
+
+    @overload
+    def __init__(
+        self: RaisesGroup[E2 | BaseExceptionGroup[E3]],
+        exception: type[E2] | Matcher[E2] | RaisesGroup[E3],
+        *other_exceptions: type[E2] | Matcher[E2] | RaisesGroup[E3],
+        match: str | Pattern[str] | None = None,
+        check: (
+            Callable[[BaseExceptionGroup[E2 | BaseExceptionGroup[E3]]], bool] | None
+        ) = None,
+    ) -> None: ...
+
+    def __init__(
+        self: RaisesGroup[Ec2 | E2 | BaseExceptionGroup[E3]],
+        exception: type[E2] | Matcher[E2] | RaisesGroup[E3],
+        *other_exceptions: type[E2] | Matcher[E2] | RaisesGroup[E3],
         allow_unwrapped: bool = False,
         flatten_subgroups: bool = False,
         match: str | Pattern[str] | None = None,
-        check: Callable[[BaseExceptionGroup[E]], bool] | None = None,
+        # NOTE: I don't think the following argument type *should* work. But it does!
+        check: (
+            Callable[[BaseExceptionGroup[E2]], bool]
+            | Callable[[ExceptionGroup[Ec2]], bool]
+            | None
+        ) = None,
```

Comment:
hmm, why accept this? I'm trying to come up with a case we want to accept but failing

Reply:
> For the record the actual type in __init__ is only used by the code within. It's not publicly shown.

oh derp. Though then the comment confuses me even more lol

Reply:
doesn't anything "work" because you have `type: ignore[arg-type]` on the only line that calls `self.check`?

The type from the signature does get exposed in `RaisesGroup(Exception).check`, but it seems all kinds of messy right now:
```python
def check_check() -> None:
    def my_check(e: ExceptionGroup[ValueError]) -> bool:
        return True
    r = RaisesGroup(ValueError, check=my_check)
    r.check(ExceptionGroup("", [ValueError()]))
```
```
error: "None" not callable  [misc]
error: Argument 1 has incompatible type "ExceptionGroup[ValueError]"; expected "BaseExceptionGroup[BaseExcT_1]"  [arg-type]
error: List item 0 has incompatible type "ValueError"; expected "ExcT_1"  [list-item]
```

Not 100% it's possible to improve it much though, might just end up with
```python
check: Callable[[BaseExceptionGroup[BaseException]], bool] | None
```

Reply:
(Not at all computer so can't check things right now)
> doesn't anything "work" because you have `type: ignore[arg-type]` on the only line that calls `self.check`?

Not exactly. It still needs to be a subtype of every overload's argument for that (which was why I was surprised and left the comment)

> 
> 
> 
> The type from the signature does get exposed in `RaisesGroup(Exception).check`, but it seems all kinds of messy right now:
> 
> ```python
> 
> def check_check() -> None:
> 
>     def my_check(e: ExceptionGroup[ValueError]) -> bool:
> 
>         return True
> 
>     r = RaisesGroup(ValueError, check=my_check)
> 
>     r.check(ExceptionGroup("", [ValueError()]))
> 
> ```
> 
> ```
> 
> error: "None" not callable  [misc]
> 
> error: Argument 1 has incompatible type "ExceptionGroup[ValueError]"; expected "BaseExceptionGroup[BaseExcT_1]"  [arg-type]
> 
> error: List item 0 has incompatible type "ValueError"; expected "ExcT_1"  [list-item]
> 
> ```
> 

Ouch I should improve it and add a test

> 
> 
> Not 100% it's possible to improve it much though, might just end up with
> 
> ```python
> 
> check: Callable[[BaseExceptionGroup[BaseException]], bool] | None
> 
> ```

I don't think this specific one works because it's not a subtype of every overload's argument, because callable is covariant in its arguments. Ie something that works for a specific type (like ValueError) might not work in general (for BaseException). For instance their check function may use an attribute on the exception.

However I agree that in general there's probably a better type hint. I'll look at it a bit.

Reply:
OK I think the issue with `check` is a mypy bug:
```py
from __future__ import annotations

from typing import Generic, TypeVar, overload, Callable

T = TypeVar("T")
U1 = TypeVar("U1")
U2 = TypeVar("U2")

class X(Generic[T]):
    @overload
    def __init__(self: X[U1], *, x: U1) -> None: ...
    @overload
    def __init__(self: X[U2], *, y: U2) -> None: ...
    
    def __init__(self: X[U1 | U2], *, x: U1 | None = None, y: U2 | None = None, z: Callable[[U1], None] | Callable[[U2], None] | None = None):
        self.z = z

reveal_type(X(x=5).z)  # N: Revealed type is "Union[def (U1`-1), def (U2`-2), None]"
```

Ironically pyright has a bug in the same exact way :P

I guess I won't be fixing the typing. FWIW the type error you got from calling `.check` was because it's possible optional. It's still wrong because it has references to the typevars and not concrete classes of course.

Reply:
If you made an issue report for mypy, could you link it here if possible?

Reply:
I made https://github.com/python/mypy/issues/18185